src/execution/operator/join/physical_iejoin.cpp

```diff
@@ -0,0 +1,1415 @@
+#include "duckdb/execution/operator/join/physical_iejoin.hpp"
+
+#include "duckdb/common/fast_mem.hpp"
+#include "duckdb/common/operator/comparison_operators.hpp"
+#include "duckdb/common/row_operations/row_operations.hpp"
+#include "duckdb/common/sort/comparators.hpp"
+#include "duckdb/common/sort/sort.hpp"
+#include "duckdb/common/vector_operations/vector_operations.hpp"
+#include "duckdb/execution/expression_executor.hpp"
+#include "duckdb/main/client_context.hpp"
+#include "duckdb/parallel/event.hpp"
+#include "duckdb/parallel/thread_context.hpp"
+
+namespace duckdb {
+
+PhysicalIEJoin::PhysicalIEJoin(LogicalOperator &op, unique_ptr<PhysicalOperator> left,
+                               unique_ptr<PhysicalOperator> right, vector<JoinCondition> cond, JoinType join_type,
+                               idx_t estimated_cardinality)
+    : PhysicalComparisonJoin(op, PhysicalOperatorType::IE_JOIN, move(cond), join_type, estimated_cardinality) {
+	// Reorder the conditions so that ranges are at the front.
+	// TODO: use stats to improve the choice?
+	// TODO: Prefer fixed length types?
+	auto conditions_p = std::move(conditions);
+	conditions.resize(conditions_p.size());
+	idx_t range_position = 0;
+	idx_t other_position = conditions_p.size();
+	for (idx_t i = 0; i < conditions_p.size(); ++i) {
+		switch (conditions_p[i].comparison) {
+		case ExpressionType::COMPARE_LESSTHAN:
+		case ExpressionType::COMPARE_LESSTHANOREQUALTO:
+		case ExpressionType::COMPARE_GREATERTHAN:
+		case ExpressionType::COMPARE_GREATERTHANOREQUALTO:
+			conditions[range_position++] = std::move(conditions_p[i]);
+			break;
+		case ExpressionType::COMPARE_NOTEQUAL:
+		case ExpressionType::COMPARE_DISTINCT_FROM:
+			// Allowed in multi-predicate joins, but can't be first/sort.
+			conditions[--other_position] = std::move(conditions_p[i]);
```

Comment:
This appears to be untested

Reply:
Ah, there is a benchmark that does this, but I'll add a small version of Q2 (I still don't have a decent generator for their Events table...)

Reply:
Heh - it has to be a small test because the unoptimised query melts down at 30KR(!)